### PR TITLE
[CPU] Propagate "scalability" when using peeling for vectorisation

### PR DESCRIPTION
**Pull request by banach-space**
Adds the missing logic to enable scalable vectorisation when using
`CPUDoubleTilingPeelingExpert` (i.e. peeling). This boils down to making
sure that "scalable" flags are correctly propagated when using
"peeling".

While the default strategy for SVE remains "masking", we might change
that in the near future to avoid calculating masks where possible.

Also, additional logging is added to aid debugging. This is mostly to
make sure that similar level of outout is generated for both "masking"
and "peeling".

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/CMakeLists.txt [hunk 1]
```diff
@@ -41,6 +41,7 @@ iree_lit_test_suite(
     "pipeline_vectorize_nd_extract_tests.mlir"
     "scalable_tile_and_vectorize_matmul.mlir"
     "select_aarch64_lowering_strategy.mlir"
+    "select_aarch64_lowering_strategy_matmul_peeling.mlir"
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/CMakeLists.txt]: [optional] I think it is better to categorize files with features but not op types.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 2]
```diff
@@ -778,8 +776,42 @@ static LogicalResult setMatmulPeelingRootConfig(
...
+  }
+  ScalableTileFlagsListType newScalableTileFlags;
+  // No scalable:
+  // * distribution,
+  // * cache parallel, and
+  // * cache reduction
+  // tile sizes.
+  newScalableTileFlags.emplace_back(numTilingDims, false);
+  newScalableTileFlags.emplace_back(numTilingDims, false);
+  newScalableTileFlags.emplace_back(numTilingDims, false);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:839]: Out of curiosity, does scalable tile flags can only be true at vector level?

I'm thinking other mechanisms about setting tile sizes, and the answer could motivate me to prioritize the work.

**Comment by dcaballe** [comment]
> The default value for -iree-codegen-enable-vector-peeling is updated
to false. 

The existing flag is more about "allow peeling in the optimization strategies when the heuristics decide to use it" than enforcing peeling as the only strategy. I think we should rename/improve the doc for this flag. I also like the idea of having a flag to enforce one strategy. As you said, having that for testing would be really good. Can we go with the extra flag approach then?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_lowering_strategy_matmul_peeling.mlir [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        %3 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [15, 14], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<15x14xf32>> -> tensor<15x14xf32>
+        %4 = flow.dispatch.tensor.load %1, offsets = [0, 0], sizes = [14, 7], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<14x7xf32>> -> tensor<14x7xf32>
+        %5 = flow.dispatch.tensor.load %2, offsets = [0, 0], sizes = [15, 7], strides = [1, 1] : !flow.dispatch.tensor<readwrite:tensor<15x7xf32>> -> tensor<15x7xf32>
+        %6 = linalg.matmul ins(%3, %4 : tensor<15x14xf32>, tensor<14x7xf32>) outs(%5 : tensor<15x7xf32>) -> tensor<15x7xf32>
+        flow.dispatch.tensor.store %6, %2, offsets = [0, 0], sizes = [15, 7], strides = [1, 1] : tensor<15x7xf32> -> !flow.dispatch.tensor<readwrite:tensor<15x7xf32>> return }
+    }
+  }
+}
+
+// TODO: FIXME
```
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_lowering_strategy_matmul_peeling.mlir]: Would be good to clarify what needs fixing here

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_lowering_strategy_matmul_peeling.mlir [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        %5 = flow.dispatch.tensor.load %2, offsets = [0, 0], sizes = [1, 1], strides = [1, 1] : !flow.dispatch.tensor<readwrite:tensor<1x1xf32>> -> tensor<1x1xf32>
+        %6 = linalg.matmul ins(%3, %4 : tensor<1x1xf32>, tensor<1x1xf32>) outs(%5 : tensor<1x1xf32>) -> tensor<1x1xf32>
+        flow.dispatch.tensor.store %6, %2, offsets = [0, 0], sizes = [1, 1], strides = [1, 1] : tensor<1x1xf32> -> !flow.dispatch.tensor<readwrite:tensor<1x1xf32>>
+        return
+      }
+    }
+  }
+}
+
+// TODO: FIXME
```
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_lowering_strategy_matmul_peeling.mlir]: Same here :)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 5]
```diff
@@ -1136,24 +1168,29 @@ setRootConfig(func::FuncOp entryPointFn,
...
   LLVM_DEBUG(KD_DBGS() << "Cache tile sizes: " << cacheTileSizes << "\n");
   LLVM_DEBUG(KD_DBGS() << "Vector tile sizes: " << vecTileSizes << "\n");
+  LLVM_DEBUG(KD_DBGS() << "Vector scalable tile sizes: " << vecTileSizes
+                       << "\n");
   LLVM_DEBUG(KD_DBGS() << "Vector scalable tile flags: " << vecScalableFlags
                        << "\n");
   LLVM_DEBUG(KD_DBGS() << "Vector size: " << vectorSize << "\n");
 
   if (usePeelingPipeline) {
     // TODO: Use scalable vector sizes.
```
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: This TODO is now done, I think :)

**Review by MacDue — APPROVED** [review]
LGTM, few little comments

**Comment by banach-space** [comment]
> The existing flag is more about "allow peeling in the optimization strategies when the heuristics decide to use it" than enforcing peeling as the only strategy. I think we should rename/improve the doc for this flag.

OK, thanks for the clarification!

> I also like the idea of having a flag to enforce one strategy. As you said, having that for testing would be really good. Can we go with the extra flag approach then?

Yup, I will be sending an update shortly. In general, it sounds like we want to consider two cases:
* the user specifies the pre-processing strategy with a CL flag and that should be respected,
* when no flag is used, we should rely on the compiler heuristics.

And then, on top of that, we'd have something specifically for peeling.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_lowering_strategy_matmul_peeling.mlir (hunk 3 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_lowering_strategy_matmul_peeling.mlir]: Fixed in the latest revision

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/CMakeLists.txt (hunk 1 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/CMakeLists.txt]: Good point, but right now I am only testing (and working on) `linalg.matmul`. A test file with a more generic name would suggest that other ops are also supported and that's not the case. Not yet. Not something I feel strongly about though!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 2 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:839]: > Out of curiosity, does scalable tile flags can only be true at vector level?

Tl;Dr That would be a safe assumption. 

Basically, that's how we use them today and I can't really think of a case in which we'd want to change that. Fundamentally, "scalability" is a concept only really available at the vector level, i.e. only the actual vector registers are scalable.

So yes, there might be scope for simplifying things.

**Comment by dcaballe** [comment]
> the user specifies the pre-processing strategy with a CL flag and that should be respected,

I guess the problem is... to which ops should the user strategy be applied? Mostly because not all the ops support all the strategies and not all the ops may perform well on the same strategy. Not sure if we can have something that is powerful enough to use beyond testing but would definitely like to hear what you have in mind!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/CMakeLists.txt (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/CMakeLists.txt]: I think thats ok. Make sense to drop the "matmul" qualifier. But maybe useful to add "scalable_vectors` qualifiers...

**Review by MaheshRavishankar — COMMENTED** [review]
This PR seems to be doing two things... setting configuration for peeling strategy and also disabling peeling through command line. Does scalable vector work  prefer the masking strategy?

**Comment by banach-space** [comment]
> Not sure if we can have something that is powerful enough to use beyond testing but would definitely like to hear what you have in mind!

Indeed and that's basically what I had in mind - the flags would only be used to facilitate testing.

**Comment by banach-space** [comment]
> This PR seems to be doing two things... setting configuration for peeling strategy and also disabling peeling through command line. 

Yeah, it's grown a bit. I will split it into two.

> Does scalable vector work prefer the masking strategy?

Masking comes at a price and is better avoided. Peeling with the remainder loop masked is going to perform much better. However, I need a few more changes before I can make the switch.

**Comment by MaheshRavishankar** [comment]
> > This PR seems to be doing two things... setting configuration for peeling strategy and also disabling peeling through command line.
> 
> Yeah, it's grown a bit. I will split it into two.
> 
> > Does scalable vector work prefer the masking strategy?
> 
> Masking comes at a price and is better avoided. Peeling with the remainder loop masked is going to perform much better. However, I need a few more changes before I can make the switch.

Thats interesting. It would be very informative to see some numbers if you can share them (if and when you have them. I am just curious).

**Comment by banach-space** [comment]
> Thats interesting. It would be very informative to see some numbers if you can share them (if and when you have them. I am just curious).

This will obviously be quite nuanced (i.e. depend on the microarchitecture etc), but you can check [Arm® Neoverse™ V1 Software Optimization Guide](https://developer.arm.com/documentation/PJDOC-466751330-9685/0101/) and compare throughput for:
* [whilelt](https://developer.arm.com/documentation/ddi0596/2021-12/SVE-Instructions/WHILELT--While-incrementing-signed-scalar-less-than-scalar-?lang=en) (predicated loop), and
* [cmp](https://developer.arm.com/documentation/ddi0602/2023-03/Base-Instructions/CMP--immediate---Compare--immediate---an-alias-of-SUBS--immediate--?lang=en) - unpredictable loop.

Now, this happens to be micro-arch that I am benchmarking on and so far peeling looks quite promising. This might (and likely will be) different for other implementations. Long term we are interested in a framework that allows us to easily switch between different approaches. So I care about both Masking and Peeling. However, ATM I am focusing on a best possible performance for a particular implementation.

Btw, I will be posting some numbers (at least relative) when updating the IREE config.

**Comment by banach-space** [comment]
I've extracted the parts introducing the new CL option to https://github.com/openxla/iree/pull/16078 and updated the summary.

**NOTE** Don't review the top commit (that belongs to #16078). And please let me know if there's a better approach for stacked PRs in IREE :)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 6]
```diff
@@ -785,7 +788,11 @@ static LogicalResult setMatmulPeelingRootConfig(
   for (const auto &[index, size] : llvm::enumerate(distTileSizes)) {
     if (!size)
       continue;
-    parallelTileSizes[index] = std::min(parallelTileSizes[index], size);
+    // The backend struggles to legalize non-power-of-two scalable vectors,
+    // hence the extra rounding up.
+    parallelTileSizes[index] =
+        roundUpToPow2(std::min(parallelTileSizes[index], size),
+                      /*predicate=*/vecScalableDims[index]);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Do we need this if we land https://github.com/openxla/iree/pull/16041 ?

Based on  the function name, this is used by peeling expert. We should not model any logics about masking here. So I think we can move it to below and make it:

```
// The backend struggles to legalize non-power-of-two scalable vectors,
// hence the extra rounding up.
if (hasSVE...) {
  for (auto &size : parallelTileSizes) {
    size = roundUpToPow2(size, /*predicate=*/vecScalableDims[index]);
  }
}
splitParallelAndReductionTiles(...);
```

I haven't landed my PR because there are regressions about RISC-V. And I'd like to get eyes from @dcaballe . I checked with him offline, and hopefully he can take a look soon. If you are +1 on my changes, maybe you can rebase on it and I can review this commit again.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 2 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:839]: SGTM! I will take it into account in the design.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 6 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I've just rebased - please check the top patch and let me know whether that's what you had in mind. I do have one question though - see my comment in [select_aarch64_sve_lowering_strategy_peeling.mlir](https://github.com/openxla/iree/pull/16058/commits/75d8954c550879b5d46f1d6ad638973b0e3183b0#diff-c694d8ea45616ae2fbacd6267d761426469481ec08256c1ce53cb0998dfbf2f7).

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_sve_lowering_strategy_peeling.mlir [hunk 7]
```diff
@@ -91,7 +87,7 @@ hal.executable private @matmul_static_tensors_sve  {
   }
 }
 
-//   CHECK-DAG: #[[CONFIG:.+]] = #iree_codegen.lowering_config<tile_sizes = {{\[}}[5, 7, 0], [5, 7, 0], [0, 0, 0], [5, 7, 0], [0, 0, 1], [0, 0, 0]]>
+//   CHECK-DAG: #[[CONFIG:.+]] = #iree_codegen.lowering_config<tile_sizes = {{\[}}[5, 7, 0], [5, 7, 0], [0, 0, 0], [8, [16], 0], [0, 0, 1], [0, 0, 0]]>
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_sve_lowering_strategy_peeling.mlir]: @hanhanW  https://github.com/openxla/iree/pull/16041 removes "clamping" and hence distribution and vector tile sizes are completely out of sync here. This is a pretty forced example (i.e. unlikely to happen in practice), but I do wonder - what should be the relations between two levels of tiling. Also, perhaps @dcaballe ?

File: runtime/src/iree/hal/drivers/vulkan/dynamic_symbol_tables.h [hunk 8]
```diff
@@ -319,7 +319,7 @@ namespace vulkan {
   INS_PFN(EXCLUDED, vkGetDisplayPlaneCapabilitiesKHR)                   \
   INS_PFN(EXCLUDED, vkGetDisplayPlaneSupportedDisplaysKHR)              \
   INS_PFN(OPTIONAL, vkGetPhysicalDeviceCalibrateableTimeDomainsEXT)     \
-  INS_PFN(OPTIONAL, vkGetPhysicalDeviceCooperativeMatrixPropertiesKHR)  \
+  INS_PFN(EXCLUDED, vkGetPhysicalDeviceCooperativeMatrixPropertiesNV)   \
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/vulkan/dynamic_symbol_tables.h]: Were these Vulkan changes intended to be included here? They seem unrelated - some git accident?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_sve_lowering_strategy_peeling.mlir (hunk 7 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_sve_lowering_strategy_peeling.mlir]: Diego and I had an offline discussion about distribution tile sizes before. It's ok if there is a bit of load imbalance. We just have to make sure that the gap between the first N-1 and the last one is not huge. We want to make distribution tile sizes be multiple of the vector length. It makes sure that we are not penalizing all the threads.

**Comment by hanhanW** [comment]
I think your rebase is broken. Maybe you can try 

```
git rebase -i main # Only choose your commit
wget https://patch-diff.githubusercontent.com/raw/openxla/iree/pull/16041.patch -O /tmp/patch
git apply /tmp/patch
git commit -m "cherry-pick https://github.com/openxla/iree/pull/16041"
git rebase -i main # and swap two commits
```

**Comment by hanhanW** [comment]
> I think your rebase is broken. Maybe you can try
> 
> ```
> git rebase -i main # Only choose your commit
> wget https://patch-diff.githubusercontent.com/raw/openxla/iree/pull/16041.patch -O /tmp/patch
> git apply /tmp/patch
> git commit -m "cherry-pick https://github.com/openxla/iree/pull/16041"
> git rebase -i main # and swap two commits
> ```

hmm this might fix conflicts in a bad way. Maybe something like:

```
git show . /tmp/patch # create a patch file for your change
git checkout main
git checkout -B your-branch-name # This reset your branch head to main
wget https://patch-diff.githubusercontent.com/raw/openxla/iree/pull/16041.patch -O /tmp/patch-16041
git apply /tmp/patch-16041
git commit -m "cherry-pick https://github.com/openxla/iree/pull/16041"

git apply /tmp/patch # and fix conflicts
git commit # write your commit message

# or just use git cherry-pick your commits.
```

File: runtime/src/iree/hal/drivers/vulkan/dynamic_symbol_tables.h (hunk 8 above)
Comment (banach-space) [reply runtime/src/iree/hal/drivers/vulkan/dynamic_symbol_tables.h]: Rebase failure, sorry :(

**Comment by banach-space** [comment]
> > I think your rebase is broken. Maybe you can try
> > ```
> > git rebase -i main # Only choose your commit
> > wget https://patch-diff.githubusercontent.com/raw/openxla/iree/pull/16041.patch -O /tmp/patch
> > git apply /tmp/patch
> > git commit -m "cherry-pick https://github.com/openxla/iree/pull/16041"
> > git rebase -i main # and swap two commits
> > ```
> 
> hmm this might fix conflicts in a bad way. Maybe something like:
> 
> ```
> git show . /tmp/patch # create a patch file for your change
> git checkout main
> git checkout -B your-branch-name # This reset your branch head to main
> wget https://patch-diff.githubusercontent.com/raw/openxla/iree/pull/16041.patch -O /tmp/patch-16041
> git apply /tmp/patch-16041
> git commit -m "cherry-pick https://github.com/openxla/iree/pull/16041"
> 
> git apply /tmp/patch # and fix conflicts
> git commit # write your commit message
> 
> # or just use git cherry-pick your commits.
> ```

That's a neat trick, thanks!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 9]
```diff
@@ -775,45 +775,77 @@ static SmallVector<int64_t> getDefaultMatmulCacheSizes(linalg::LinalgOp op,
...
+  std::swap(vectorParallelTileSizes.back(), vectorReductionTileSizes.back());
+
   TileSizesListType tileSizes = {
       SmallVector<int64_t>(distTileSizes), cacheParallelTileSizes,
-      cacheReductionTileSizes, parallelTileSizes, reductionTileSizes};
-
+      cacheReductionTileSizes, vectorParallelTileSizes,
+      vectorReductionTileSizes};
   // No need for tiling inner parallel dims.
   tileSizes.emplace_back(numTilingDims, 0);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: We should move these to the end of function. Otherwise the vectorParallelTileSizes adjustment is ignored.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 10]
```diff
@@ -775,45 +775,77 @@ static SmallVector<int64_t> getDefaultMatmulCacheSizes(linalg::LinalgOp op,
...
+      continue;
+    vectorParallelTileSizes[index] =
+        roundUpToPow2(size,
+                      /*predicate=*/vecScalableDims[index]);
+  }
+
+  // 2. Set scalable flags for all the tiling levels.
+  SmallVector<bool> parallelScalableFlags;
+  for (auto [index, tileSize] : llvm::enumerate(vecTileSizes))
+    parallelScalableFlags.push_back(vecScalableDims[index]);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I think this can just be

```
SmallVector<bool> parallelScalableFlags = vecScalableDims;
```

Did I miss something?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 11]
```diff
@@ -1170,24 +1202,28 @@ setRootConfig(func::FuncOp entryPointFn,
...
   LLVM_DEBUG(KD_DBGS() << "Vector size: " << vectorSize << "\n");
 
   if (usePeelingPipeline) {
-    // TODO: Use scalable vector sizes.
-    return setMatmulPeelingRootConfig(entryPointFn, contractionOp,
-                                      distTileSizes, cacheTileSizes,
-                                      vecTileSizes, vectorSize);
+    return setMatmulPeelingRootConfig(
+        entryPointFn, contractionOp, distTileSizes, cacheTileSizes,
+        vecScalableFlags, vecTileSizes, vectorSize);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1226]: > Fundamentally, "scalability" is a concept only really available at the vector level, i.e. only the actual vector registers are scalable.

Based on the reply from other comments, we don't need to declare `distScalableTileFlags` and `ScalableTileFlagsListType`. We can directly pass `vecScalableFlags` to the function. This saves us a level of access (like l.805 -- `const SmallVectorImpl<bool> &vecScalableDims = inputScalableTileFlags.back();`).

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 11 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1226]: You are right and that's what I do for `setMatmulPeelingRootConfig`. However, there's also `setMatmulRootConfig` that takes both for now. I was planning to simplify that in a follow-up patch (trying to avoid making too many changes at once).

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 10 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: That particular constructor is `delete`d, but you are right that this can and should be simplified :)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 9 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Good catch, thanks!

I will move it a few lines further down - still want to keep the set-up for scalable flags separate.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 11 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1226]: In this case, maybe we can pass `inputScalableTileFlags.back()` with ArrayRef<bool> type to `setMatmulPeelingRootConfig`? It helps make the logic in `setMatmulPeelingRootConfig` easier, IMHO.

Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1226]: Agreed.

Updated  - roughly as you suggested (sorry, wrote the code before saw your comment).

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 12]
```diff
@@ -775,7 +775,7 @@ static SmallVector<int64_t> getDefaultMatmulCacheSizes(linalg::LinalgOp op,
 static LogicalResult setMatmulPeelingRootConfig(
     func::FuncOp entryPointFn, linalg::ContractionOpInterface op,
     ArrayRef<int64_t> distTileSizes, ArrayRef<int64_t> cacheTileSizes,
-    const ScalableTileFlagsListTypeRef inputScalableTileFlags,
+    SmallVector<bool> &inputVecScalableTileFlags,
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Should this be `ArrayRef<bool>` type? Otherwise we could accidentally update the data.

**Review by hanhanW — APPROVED** [review]
LGTM, let's wait until landing the parent PRs. Thanks for pushing on this!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 13]
```diff
@@ -801,15 +804,59 @@ static LogicalResult setMatmulPeelingRootConfig(
...
 
+  // 2. Set scalable flags for all the tiling levels.
+  SmallVector<bool> parallelScalableFlags(inputVecScalableTileFlags.begin(),
+                                          inputVecScalableTileFlags.end());
+  SmallVector<bool> reductionScalableFlags(numTilingDims, false);
+
+  // Ensure there's no zero scalable dims.
+  for (unsigned i = 0; i < numTilingDims; i++) {
+    if (vectorReductionTileSizes[i] == 0)
+      reductionScalableFlags[i] = false;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Do we need this? It looks like reductionScalableFlags[*] is always false.

We are assuming it is a `[B], M, N, K` kind of matmul, (which is actually not good, but probably fine for now). How about following the logic like other tile sizes? E.g.,

1. we copy `inputVecScalableTileFlags` to `parallelScalableFlags`.
2. initialize `reductionScalableFlags` to all false.
3. swap the last element for `parallelScalableFlags` and `reductionScalableFlags`.

Does it work (and look better)?

```cpp
  SmallVector<bool> parallelScalableFlags(inputVecScalableTileFlags.begin(),
                                          inputVecScalableTileFlags.end());
  SmallVector<bool> reductionScalableFlags(numTilingDims, false);
  std::swap(parallelScalableFlags.back(), reductionScalableFlags.back());
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 14]
```diff
@@ -801,15 +804,59 @@ static LogicalResult setMatmulPeelingRootConfig(
...
 
+  // The backend struggles to legalize non-power-of-two scalable vectors,
+  // hence the extra rounding up.
+  for (const auto &[index, size] : llvm::enumerate(vectorParallelTileSizes)) {
+    if (!size)
+      continue;
+    vectorParallelTileSizes[index] =
+        roundUpToPow2(size,
+                      /*predicate=*/inputVecScalableTileFlags[index]);
+  }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Do we need the adjustment for reduction tile size? How about moving this to l.801 (before setting tile sizes for `vectorParallelTileSizes` and `vectorReductionTileSizes`), and apply it on `clampedVecTileSizes`?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 13 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Yup, that's much nicer, thanks! Sending update version shortly.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 14 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: > Do we need the adjustment for reduction tile size?

We don't use scalable vectors for the reduction dim, but we might do in do feature. In any case, it will be safer to round up both sets, thanks!

> How about moving this to l.801 (before setting tile sizes for vectorParallelTileSizes and vectorReductionTileSizes), and apply it on clampedVecTileSizes?

I'll move it even further up and keep all the pre-processing for scalable vectors together.

**Review by hanhanW — APPROVED** [review]
LGTM, thanks you!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 15]
```diff
@@ -787,6 +790,17 @@ static LogicalResult setMatmulPeelingRootConfig(
     clampedVecTileSizes[index] = std::min(clampedVecTileSizes[index], size);
   }
 
+  // The LLVM backend struggles to legalize non-power-of-two scalable vectors,
+  // hence the extra rounding up.
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:794]: I would say this is true for all the backends. Do we have cases where we generate non power of two for the main vector dimension? 

We also have to make sure we don't round up sizes of unrolled dimensions. It should be fine to unroll by 7, for example.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:794]: I wonder that if this makes sense when the dim size is smaller than vector size? @banach-space mentioned that the "clamping" is needed in scalable vectorization. I think he is evaluating if it is really needed or not.

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:794]: I think this was originally my comment, what I meant here was fixed vectors can always be scalarized (though may generate poor code). Scalable vectors cannot always be scalarized/legalized, so can result in the backend crashing. E.g.  `vector<[7]xf32>`, probably can't be legalized.

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:794]: >  Do we have cases where we generate non power of two for the main vector dimension?

It did used to happen for statically sized tensors.

Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:794]: > Do we have cases where we generate non power of two for the main vector dimension?

In the past I've noticed that happening due to "clamping" - that's one reason I'd like to remove it.

> I wonder that if this makes sense when the dim size is smaller than vector size? @banach-space mentioned that the "clamping" is needed in scalable vectorization. I think he is evaluating if it is really needed or not.

I will prioritise that next week 👍🏻

Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:794]: I think this is a legacy decision that we made in an attempt to make the inner tile sizes multiple of the distribution sizes but I think it doesn't hold anymore. When we use masking, we want to always round up to the nearest power of two. However, when we use peeling, we want to make sure the main vector loop executes at least once so we should round up to the nearest power of two only if there are enough iterations for that to happen (or an unknown number of iterations). Otherwise, we should round down to the nearest power of two and let the remainder kick in.
